pkg/network/ebpf/c/protocol-classification-helpers.h

```diff
@@ -0,0 +1,214 @@
+static __always_inline bool is_http(const char *buf, __u32 size) {
+    if (size < HTTP_MIN_SIZE) {
+        return false;
+    }
+    if (buf == NULL) {
+        return false;
+    }
+
+    if ((buf[0] == 'H') && (buf[1] == 'T') && (buf[2] == 'T') && (buf[3] == 'P')) {
+        return true;
+    } else if ((buf[0] == 'G') && (buf[1] == 'E') && (buf[2] == 'T') && (buf[3]  == ' ') && (buf[4] == '/')) {
+        return true;
+    } else if ((buf[0] == 'P') && (buf[1] == 'O') && (buf[2] == 'S') && (buf[3] == 'T') && (buf[4]  == ' ') && (buf[5] == '/')) {
+        return true;
+    } else if ((buf[0] == 'P') && (buf[1] == 'U') && (buf[2] == 'T') && (buf[3]  == ' ') && (buf[4] == '/')) {
+        return true;
+    } else if ((buf[0] == 'D') && (buf[1] == 'E') && (buf[2] == 'L') && (buf[3] == 'E') && (buf[4] == 'T') && (buf[5] == 'E') && (buf[6]  == ' ') && (buf[7] == '/')) {
+        return true;
+    } else if ((buf[0] == 'H') && (buf[1] == 'E') && (buf[2] == 'A') && (buf[3] == 'D') && (buf[4]  == ' ') && (buf[5] == '/')) {
+        return true;
+    } else if ((buf[0] == 'O') && (buf[1] == 'P') && (buf[2] == 'T') && (buf[3] == 'I') && (buf[4] == 'O') && (buf[5] == 'N') && (buf[6] == 'S') && (buf[7]  == ' ') && ((buf[8] == '/') || (buf[8] == '*'))) {
+        return true;
+    } else if ((buf[0] == 'P') && (buf[1] == 'A') && (buf[2] == 'T') && (buf[3] == 'C') && (buf[4] == 'H') && (buf[5]  == ' ') && (buf[6] == '/')) {
+        return true;
+    }
+
+    return false;
+}
+
+// Determines the protocols of the given buffer. If we already classified the payload (a.k.a protocol out param
+// has a known protocol), then we do nothing.
+static __always_inline void classify_protocol(protocol_t *protocol, const char *buf, __u32 size) {
+    if (protocol == NULL || (*protocol != PROTOCOL_UNKNOWN && *protocol != PROTOCOL_UNCLASSIFIED)) {
+        return;
+    }
+
+    if (is_http(buf, size)) {
+        *protocol = PROTOCOL_HTTP;
+    } else if (is_http2(buf, size)) {
+        *protocol = PROTOCOL_HTTP2;
+    } else {
+        *protocol = PROTOCOL_UNKNOWN;
+    }
+
+    log_debug("[protocol classification]: Classified protocol as %d %d %s\n", *protocol, size, buf);
+}
+
+// Decides if the protocol_classifier should process the packet. We process not empty TCP packets.
+static __always_inline bool should_process_packet(struct __sk_buff *skb, skb_info_t *skb_info, conn_tuple_t *tup) {
+    // we're only interested in TCP traffic
+    if (!(tup->metadata&CONN_TYPE_TCP)) {
+        return false;
+    }
+
+    bool empty_payload = skb_info->data_off == skb->len;
+    return !empty_payload;
+}
+
+// The method is used to read the data buffer from the __sk_buf struct. Similar implementation as `read_into_buffer_skb`
+// from http parsing, but uses a different constant (CLASSIFICATION_MAX_BUFFER).
+static __always_inline void read_into_buffer_skb_post_4_5_0(char *buffer, struct __sk_buff *skb, skb_info_t *info) {
```

Comment:
since we don't have loops, you must know the exact number of iterations in the function, since `CLASSIFICATION_MAX_BUFFER` is different than the size in HTTP parsing, we cannot reuse the function

changed the name

Reply:
good point re: verifier